main/src/cgeo/geocaching/ui/CalculateButton.java

```diff
@@ -0,0 +1,259 @@
+
+    // The buttons own state is stored in a 'state' object to facilitate easy saving and restoring.
+    private ButtonData buttonData;
+
+    public CalculateButton(Context context) {
+        super(context);
+        setup();
+    }
+
+    public CalculateButton(Context context, AttributeSet attrs) {
+        super(context, attrs);
+        setup();
+    }
+
+    private void setup() {
+        setData(new ButtonData());
+        butt.setOnClickListener(new CoordDigitClickListener());
+    }
+
+    // The 'Label' is the 'name' that is to be displayed on the button
+    public char getLabel() {
+        char rv;
+        switch (getType()) {
+            case INPUT_VAL: rv = buttonData.inputVal;   break;
+            case AUTO_CHAR: rv = buttonData.autoChar;   break;
+            case CUSTOM:    rv = buttonData.customChar; break;
+            default:        rv='*'; // Should never happen.
+        }
+
+        return rv;
+    }
+
+    public void setLabel(char theLabel) {
+        butt.setText("" + theLabel);
+    }
+
+    public ValueType getType() {
+        return buttonData.type;
+    }
+
+    public void setType(ValueType theType) {
+        buttonData.type = theType;
+        setBackgroundColour();
+    }
+
+    // Buttons displaying a custom value are given a 'steel-like' appearance so as to distinguish them from regular buttons.
+    private  void setBackgroundColour() {
+        if (buttonData.type == ValueType.CUSTOM) {
+            butt.setBackgroundColor(ContextCompat.getColor(getContext(), R.color.button_default));
+            butt.setTextColor(ContextCompat.getColor(getContext(), R.color.steel));
+            butt.setTypeface(null, Typeface.BOLD);
+        } else {
+            butt.setBackgroundResource(R.drawable.button_background_kitkat);
+            butt.setTextColor(Color.WHITE);
+            butt.setTypeface(null, Typeface.NORMAL);
+        }
+    }
+
+    // Data used to preserve this buttons state between sessions.
+    public ButtonData getData() {
+        return buttonData;
+    }
+
+    public void setData(ButtonData buttonData) {
+        this.buttonData = buttonData;
+        setBackgroundColour();
+        updateButtonText();
+    }
+
+    public void setInputVal(String theVal) {
+        setInputVal(theVal.charAt(0));
+    }
+
+    public void setInputVal(char theVal) {
+        buttonData.inputVal = theVal;
+        updateButtonText();
+    }
+
+    public CalculateButton getNextButton() {
+        return nextButton;
+    }
+
+    public void setNextButton(CalculateButton theNext) {
+        nextButton = theNext;
+    }
+
+    /**
+     * This method is used to assign buttons alphabetically increasing names.
+     * @param currentChar:  values indicated that the autoChar will be and is updated appropriately for the next button based on the current button's state.
+     */
+    public void setAutoChar(char currentChar) {
+        if (currentChar < 'A')
+            currentChar = 'A';
```

Comment:
Indeed, please introduce a final local variable which is initialized to either 'A', 'Z' or `currentChar`.